crates/solver/src/settlement_rater.rs

```diff
@@ -244,18 +248,251 @@ impl SettlementRating for SettlementRater {
         };
 
+        // recalculate score if success probability is provided
+        let score = match settlement.success_probability {
+            Some(success_probability) => {
+                match self.score_calculator.compute_score(
+                    &objective_value,
+                    &inputs.gas_cost,
+                    success_probability,
+                ) {
+                    Ok(score) => score,
+                    Err(err) => {
+                        tracing::warn!(?err, "Failed to compute score with success probability");
+                        score
+                    }
+                }
+            }
+            None => score,
+        };
+
+        let gas_cost = match settlement.gas_cost.as_ref() {
+            Some(gas_cost) => GasCost::SolverEstimated(gas_cost.to_big_rational()),
+            None => {
+                let gas_cost = &simulation.gas_estimate.to_big_rational() * &effective_gas_price;
+                GasCost::ProtocolEstimated(gas_cost)
+            }
+        };
+
         let rated_settlement = RatedSettlement {
             id,
             settlement,
             surplus: inputs.surplus_given,
             earned_fees,
             solver_fees: inputs.solver_fees,
             gas_estimate: simulation.gas_estimate,
-            gas_price: effective_gas_price.clone(),
+            gas_price: effective_gas_price,
+            gas_cost,
             objective_value,
             score,
             ranking: Default::default(),
         };
         Ok(Rating::Ok(rated_settlement))
     }
 }
+
+pub struct ScoreCalculator {
+    score_cap: BigRational,
+    submission_strategies: Vec<TransactionStrategyArg>,
+}
+
+impl ScoreCalculator {
+    pub fn new(score_cap: BigRational, submission_strategies: Vec<TransactionStrategyArg>) -> Self {
+        Self {
+            score_cap,
+            submission_strategies,
+        }
+    }
+
+    pub fn cost_fail(&self, gas_cost: &BigRational) -> BigRational {
+        if self.submission_strategies == vec![TransactionStrategyArg::Flashbots] {
+            BigRational::zero()
+        } else {
+            gas_cost.clone()
+        }
+    }
+
+    pub fn compute_score(
+        &self,
+        objective_value: &BigRational,
+        gas_cost: &BigRational,
+        success_probability: f64,
+    ) -> Result<Score> {
+        ensure!(
+            (0.0..=1.0).contains(&success_probability),
+            "success probability must be between 0 and 1."
+        );
+        let success_probability = BigRational::from_float(success_probability).unwrap();
+        let cost_fail = self.cost_fail(gas_cost);
+        let optimal_score =
+            self.compute_optimal_score(objective_value.clone(), success_probability, cost_fail)?;
+        let score = big_rational_to_u256(&optimal_score).context("Invalid score.")?;
+        Ok(Score::ProtocolWithRisk(score))
+    }
+
+    fn compute_optimal_score(
+        &self,
```

Comment:
nit: you can exchange this method for a free method by replacing &self with score_cap (it doesn't use any other state from self). Given this is a pure math function (in -> out, no side effects expected) I think a free method is slightly more appropriate and makes it more ergonomic to test.

Reply:
moved out